includes/classes/shopping_cart.php

```diff
@@ -744,173 +767,200 @@ public function calculate()
-                                $freeShippingTotal += $text_letters;
-                                $freeShippingTotal += $text_words;
-                            }
-                            // calculate Product Price without Specials, Sales or Discounts
-                            $total_before_discounts += $text_letters;
-                            $total_before_discounts += $text_words;
+                        // calculate Product Price without Specials, Sales or Discounts
+                        $total_before_discounts += $attribute_price->fields['options_values_price'];
+                    } // eof: attribute price
+
+                    // adjust for downloads
+                    // adjust products price
+                    $sql = "SELECT products_attributes_id
+                            FROM " . TABLE_PRODUCTS_ATTRIBUTES_DOWNLOAD . "
+                            WHERE products_attributes_id = $attributes_id";
+                    $check_download = $db->Execute($sql, 1);
+                    if (!$check_download->EOF) {
+                        // count number of downloads
+                        $this->download_count += ($check_download->RecordCount() * $qty);
+                        // do not count download as free when set to product/download combo
+                        if ($free_shipping_applied === false && $adjust_downloads === 1 && $product['product_is_always_free_shipping'] !== '2') {
+                            $freeShippingTotal += $products_price;
+                            $this->free_shipping_item += $qty;
                         }
+                        // adjust for attributes price
+                        $freeShippingTotal += $new_attributes_price;
+                    }
 
-                        // attributes_price_factor
-                        $added_charge = 0;
-                        if ($attribute_price->fields['attributes_price_factor'] > 0) {
-                            //echo 'products_id: ' . $product->fields['products_id'] . ' Prices ' . '$chk_price: ' . $chk_price . ' $chk_special: ' . $chk_special . ' attributes_price_factor:' . $attribute_price->fields['attributes_price_factor'] . ' attributes_price_factor_offset: ' . $attribute_price->fields['attributes_price_factor_offset'] . '<br>';
-                            $added_charge = zen_get_attributes_price_factor($chk_price, $chk_special, $attribute_price->fields['attributes_price_factor'], $attribute_price->fields['attributes_price_factor_offset']);
+                    ////////////////////////////////////////////////
+                    // calculate additional attribute charges
+                    $chk_price = zen_get_products_base_price($uprid);
+                    $chk_special = zen_get_products_special_price($uprid, false);
+                    // products_options_value_text
+                    if (ATTRIBUTES_ENABLED_TEXT_PRICES === 'true' && zen_get_attributes_type($check_attribute) == PRODUCTS_OPTIONS_TYPE_TEXT) {
```

Comment:
❓ I think there's a bug here. `$check_attribute` got removed from old line 786, and was inlined as just `$attributes_id` on new line 816.

Reply:
Yes, there is!  Update forthcoming.